beater/middleware/rate_limit_middleware.go

```diff
@@ -0,0 +1,38 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package middleware
+
+import (
+	"github.com/elastic/apm-server/beater/api/ratelimit"
+	"github.com/elastic/apm-server/beater/config"
+	"github.com/elastic/apm-server/beater/request"
+)
+
+const burstMultiplier = 3
+
+// RateLimitMiddleware adds a rate limiter per IP to the request context
+func RateLimitMiddleware(cfg *config.EventRate) Middleware {
+	cache, err := ratelimit.NewLRUCache(cfg.LruSize, cfg.Limit, burstMultiplier)
```

Comment:
This is an error that occurs when creating and registering the API handlers. Such an error should not be carried to every single request that is wrapped by this middleware, but rather let the Server startup fail (or recover in whichever other way).

Reply:
Exactly, that is how it works unless I did something wrong

Reply:
🤦‍♀ right, I misread as part of the actual request handling, fine as is.